Documentation/design-docs/hw-intrinsics.md

```diff
@@ -0,0 +1,81 @@
+
+In discussing the hardware intrinsics, we refer to the target platform, such as X86 or Arm64, as the "platform" and each set of extensions that are implemented as a unit (e.g. AVX2 on X64 or Simd on Arm64) as an "ISA".
+
+There is a design document for the Arm64 intrinsics: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/arm64-intrinsics.md. It should be updated to reflect current (and ongoing) progress.
+
+## Overview
+
+The reference assemblies for the hardware intrinsics live in corefx, but all of the implementation is in the coreclr repo:
+
+* The C# implementation lives in coreclr/src/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that usually invoke themselves recursively (see Indirect invocation support below).
+
+## C# Implementation
+
+The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types. The vector types are considered platform-agnostic, though not all platforms define intrinsics on all of these types. They are:
+
+* `Vector64<T>` - A 64-bit vector of type `T`. For example, a `Vector64<int>` would hold two 32-bit integers.
+* `Vector128<T>` - A 128-bit vector of type `T`
+* `Vector256<T>` - A 256-bit vector of type `T`
+
+Note that these are generic types, which distinguishes these from native intrinsic vector types. It also somewhat complicates interop, as the runtime currently doesn't support interop for generic types.
+*** Is there an issue for this?? ***
+
+The C# declaration of a hardware intrinsic ISA class is marked with the `[Intrinsic]` attribute, and the implementations of the intrinsic methods on that class are recursive. When the VM encounters such a method, it will communicate to the JIT that this is an intrinsic method, and will also pass a `mustExpand` flag to indicate that the JIT must generate code.
+
+### Platform-agnostic vector types
+
+The vector types supported by one or more target ISAs are supported across platforms, though they extent to which operations on them are available and accelerated is dependent on the target ISA. These are the generic types: `Vector64<T>`, `Vector128<T>` and `Vector256<T>`.
+
+## JIT Implementation
+
+The bulk of the implementation work for hardware intrinsics is in the JIT.
+
+### Hardware Intrinsics Table
+
+There is a hardware intrinsics table for each platform that supports hardware intrinsics: currently `_TARGET_XARCH_` and `TARGET_ARM64_`. They live in hwintrinsiclistxarch.h and hwintrinsiclistarm64.h respectively.
+
+These tables are intended to capture information that can assist in making the implementation as data-driven as possible.
+
+### IR
+
+The hardware intrinsics nodes are generally imported as `GenTreeHWIntrinsic` nodes, with the `GT_HWIntrnsic` operator. On these nodes, the `gtHWIntrinsicId` field contains the intrinsic ID, as declared in the hardware intrinsics table.
+
+### Importation
+
+Hardware intrinsics appear in the IL as calls.
```

Comment:
It might be worth calling out that this is where the `IsSupported` checks need to happen and that importation should be returning `null` for things which aren't recognized at all and `throw PNSE` for things which are recognized but which aren't supported by the current platform.

Reply:
Maybe also worth calling out the constant folding story (both where we are now, and where we want to be)?

Reply:
I would suggest that here it would be worth referring to the section on `mustExpand`.

Can you expand upon your "where we want to be" comment?

Reply:
I believe we determined that we can't handle all forms of instructions that require constant inputs in the importer and we would eventually need to handle this later (possibly in morph or lowering).

So, I was saying it might be worth calling out that here and linking back to related issues (such as https://github.com/dotnet/coreclr/issues/17108).

Reply:
In general, we have a lot of issues that track various improvements/etc. These are things that the ARM intrinsics are likely to hit during implementation as well, so it would be nice to call those out and link back to the existing issues where applicable.

Reply:
Ah, I see. Thanks.

Reply:
> So, I was saying it might be worth calling out that here and linking back to related issues (such as #17108).

Ah! Thanks! I was actually wondering about how this case is handled.

> so it would be nice to call those out and link back to the existing issues where applicable.

I agree, would make it much easier to get up to speed with what is and is not working/missing.